### PR TITLE
GM long: add pitch-compensation

### PR DESCRIPTION
**Pull request by twilsonco**
([This](https://github.com/commaai/openpilot/pull/25469), but without model-predicted pitch, and this is entirely contained in gm/carcontroller)

# Description/justification

This PR adds pitch compensation to longitudinal control for GM cars, resulting in a large decrease in longitudinal error (versus the long plan) when pitch changes or when on an incline/decline.

Adds longitudinal pitch compensation to be used in cars that use gas/brake control, where the command specifies an amount of acceleration force rather than an amount of acceleration relative to the road. The implementation is contained in car controller, as a type of passive long feedforward, transparent to the controller. 

To allow this to account for longitudinal actuator delay to allow for smoothing of the pitch signal, future model pitch is used to create a smoothed, predictive pitch.

Though not shown here, another benefit of this improvement is that if more than `ACCEL_MAX` acceleration is required to maintain speed up a steep hill, stock will not be able to maintain speed. With this improvement that is no longer the case.

`actuators.accelPitchCompensated` is used to determine the gas command, which specifies an acceleration force. The brake command on gm specifies acceleration relative to the ground, so the `actuators.accel` value is the correct value for friction brakes, even when going downhill. 
To support this assertion, see that gas command does not match up with aEgo:
image
While brake command does match with aEgo
image
I also verified this extensively (though not with proper documentation) when developing "one pedal mode" in my openpilot fork, which involved artificially sending a constant brake command and observing the car's braking behavior. I found that the same brake command produces far more braking force on a decline, and far less (or none) on an incline.

However, when openpilot requests negative acceleration in order to stop behind a lead _when going uphill_, it would engage the friction brakes sooner (at the same time it would now, without this improvement), despite the negative gravitational acceleration, and resulting in more braking than in necessary, since it's being assisted by gravity.
So you'd like for openpilot to postpone friction brake use when decelerating uphill, but it can become a problem when stopping, as the pitch-compensated acceleration may remain positive even though openpilot would like to stop, resulting in late stopping behind leads.
The solution is to smoothly switch to non-pitch-compensated acceleration for determining brakes at low speed.
I've been using this method on my fork for about a month and it keeps the efficiency and brake-avoidance benefits of pitch-compensation without introducing possible problems when stopping behind leads.

(Any issues caused by the gas and brake signals having different physical meanings will also be minimized by [using the correct gas/brake threshold acceleration value](https://github.com/commaai/openpilot/pull/25453))

**This should also be used on other makes that specify acceleration force with gas or brake commands**

Route: c11fcb510a549332|2022-08-19--09-03-54--0

### CODE REVIEW

**Comment by twilsonco** [comment]
Route: c11fcb510a549332|2022-08-19--09-03-54--0

# Verification

Here we compare master with present and future pitch compensation on a couple steep steps of a hill, going up and down the steps.

Pitch compensation cuts speed error in half, and lowers acceleration error by 13%. 
Additionally, this eliminates the problem of `MAX_ACCEL` (2m/s^2) being insufficient for climbing steep hills; with pitch compensation op can accelerate up any hill the car can.

(In the future, when Comma has sufficiently verified the model predicted pitch, I'd like to [include future pitch in order to compensate for long actuator delay](https://github.com/commaai/openpilot/pull/25469), which, as you can see here, decreases speed and acceleration error by an additional 46% compared to when using instantaneous pitch. Compared to master, future pitch compensation reduces speed and acceleration error by 74% and 53% respectively, resulting in greater passenger comfort and improved vehicle efficiency.)

pitch compare

**Comment by sshane** [comment]
@HaraldSchafer would Bolt long be helpful in verifying this? I can try to get that merged

**Comment by sshane** [comment]
It's possible we need something similar to this for Toyotas to undo the accel compensation the PCM does, so we should try to make this as generic as possible.

**Comment by twilsonco** [comment]
So something in between this and the [previous implementation](https://github.com/commaai/openpilot/pull/25469) that was unnecessarily couched inside of liveParams?

**Comment by adeebshihadeh** [comment]
This isn't the large; blocking this on abstracting it out seems like premature optimization.

File: selfdrive/car/gm/carcontroller.py [hunk 1]
```diff
@@ -87,6 +109,9 @@ def update(self, CC, CS):
         send_fcw = hud_alert == VisualAlert.fcw
         can_sends.append(gmcan.create_acc_dashboard_command(self.packer_pt, CanBus.POWERTRAIN, CC.enabled,
                                                             hud_v_cruise * CV.MS_TO_KPH, hud_control.leadVisible, send_fcw))
+      else:
+        accel_g = ACCELERATION_DUE_TO_GRAVITY * apply_deadzone(self.pitch.x, PITCH_DEADZONE) # driving uphill is positive pitch
+        accel += accel_g
```
Comment (sshane) [inline selfdrive/car/gm/carcontroller.py:114]: any way you can remove the duplicate calculation?

File: selfdrive/car/gm/carcontroller.py [hunk 2]
```diff
@@ -60,20 +71,31 @@ def update(self, CC, CS):
 
       can_sends.append(gmcan.create_steering_control(self.packer_pt, CanBus.POWERTRAIN, apply_steer, idx, CC.latActive))
 
+
+
+
```
Comment (sshane) [inline selfdrive/car/gm/carcontroller.py:76]:
```suggestion
```

File: selfdrive/car/gm/carcontroller.py [hunk 3]
```diff
@@ -60,20 +71,31 @@ def update(self, CC, CS):
...
+    
     if self.CP.openpilotLongitudinalControl:
       # Gas/regen, brakes, and UI commands - all at 25Hz
       if self.frame % 4 == 0:
+        # Pitch compensated acceleration;
+        # TODO: include future pitch (sm['modelDataV2'].orientation.y) to account for long actuator delay
+        pitch = clip(CC.orientationNED[1], self.pitch.x - PITCH_MAX_DELTA, self.pitch.x + PITCH_MAX_DELTA)
+        pitch = clip(pitch, PITCH_MIN, PITCH_MAX)
+        self.pitch.update(pitch)
```
Comment (sshane) [inline selfdrive/car/gm/carcontroller.py:84]: Feels like the delta clip and the filter do similar things, any reason for both?

File: selfdrive/car/gm/carcontroller.py [hunk 4]
```diff
@@ -25,13 +34,15 @@ def __init__(self, dbc_name, CP, VM):
...
     self.params = CarControllerParams()
+    self.pitch = FirstOrderFilter(0., 0.09 * 4, DT_CTRL * 4) # runs at 25 Hz
 
     self.packer_pt = CANPacker(DBC[self.CP.carFingerprint]['pt'])
     self.packer_obj = CANPacker(DBC[self.CP.carFingerprint]['radar'])
     self.packer_ch = CANPacker(DBC[self.CP.carFingerprint]['chassis'])
 
   def update(self, CC, CS):
     actuators = CC.actuators
+    accel = actuators.accel
```
Comment (sshane) [inline selfdrive/car/gm/carcontroller.py:45]: use class variable so you don't need to set it when not an active control frame. check toyota/carcontroller

File: selfdrive/car/gm/carcontroller.py [hunk 5]
```diff
@@ -25,13 +34,15 @@ def __init__(self, dbc_name, CP, VM):
     self.lka_icon_status_last = (False, False)
 
     self.params = CarControllerParams()
+    self.pitch = FirstOrderFilter(0., 0.09 * 4, DT_CTRL * 4) # runs at 25 Hz
```
Comment (sshane) [inline selfdrive/car/gm/carcontroller.py:37]: if we only use the filter, ensure that going from min to max pitch doesn't violate your selected rate limit above

File: selfdrive/car/gm/carcontroller.py [hunk 6]
```diff
@@ -60,20 +71,31 @@ def update(self, CC, CS):
...
+        pitch = clip(pitch, PITCH_MIN, PITCH_MAX)
+        self.pitch.update(pitch)
+        accel_g = ACCELERATION_DUE_TO_GRAVITY * apply_deadzone(self.pitch.x, PITCH_DEADZONE) # driving uphill is positive pitch
+        accel += accel_g
         if not CC.longActive:
           # Stock ECU sends max regen when not enabled
           self.apply_gas = self.params.MAX_ACC_REGEN
           self.apply_brake = 0
         else:
+          brake_accel = actuators.accel + accel_g * interp(CS.out.vEgo, BRAKE_PITCH_FACTOR_BP, BRAKE_PITCH_FACTOR_V)
```
Comment (sshane) [inline selfdrive/car/gm/carcontroller.py:92]: if we change brake like this we probably should change gas

File: selfdrive/car/gm/carcontroller.py (hunk 1 above)
Comment (sshane) [inline selfdrive/car/gm/carcontroller.py:114]: use class variable

**Review by sshane — CHANGES_REQUESTED** [review]

**Comment by simontheflutist** [comment]
With a view towards https://github.com/commaai/openpilot/issues/30341, econd-order effects would change the calculation from `pitch` to `sin(pitch) cos(roll)`. I can't say for sure if it's necessary, but freeway ramps, which can have pronounced pitch and roll, come to mind.

**Comment by twilsonco** [comment]
Thanks @simontheflutist.

Oops forgot to close this. This too will be superseded by #29659.

**Comment by twilsonco** [comment]
Closed this in favor of #29659  which is now closed. Reopening

**Comment by twilsonco** [comment]
I'll get Shanes's proposed changes done and a new test route so that this can be merged

**Comment by sshane** [comment]
We've moved the car interfacing code to our [`opendbc`](https://github.com/commaai/opendbc) repository, which is now the new home for car ports and fingerprints. Please re-open your pull request against [`opendbc`](https://github.com/commaai/opendbc) at your convenience by using this command below. This will transform all changes under `selfdrive/car/` to `opendbc_repo/opendbc/car/`. Make sure you have initialized submodules beforehand and have checked out the latest `opendbc` commit.

```bash
PR_NUMBER=33045
curl -L https://github.com/commaai/openpilot/pull/$PR_NUMBER.patch | sed -e 's/selfdrive\/car/opendbc_repo\/opendbc\/car/g' | git apply -v --reject
```

Simply replace the PR number with your own. Once done, add the files, fix any conflicts, and open a new PR. Alternatively, you may start a new PR from scratch if that is easier for you.

**Comment by sshane** [comment]
If you would like to re-open, you can run our new [maneuver tool](https://github.com/commaai/openpilot/tree/master/tools/longitudinal_maneuvers) and share the before and after reports. Here are some examples of the report showing a clear and obvious improvement, enabling us to merge a new tune faster and easier:

https://github.com/commaai/opendbc/pull/1248

https://github.com/commaai/opendbc/pull/1260

**Pull request closed without merging**